grakn-graph/src/main/java/ai/grakn/graph/internal/AbstractGraknGraph.java

```diff
@@ -285,15 +288,105 @@ private EdgeImpl addEdge(Concept from, Concept to, Schema.EdgeLabel type){
+        Map<TypeName, TypeImpl> cloneCache = localCloneCache.get();
+        if(cloneCache == null){
+            localCloneCache.set(cloneCache = new HashMap<>());
+        }
+        return cloneCache;
+    }
+
+    /**
+     * Given a concept log bound to a specific thread the central ontology cache is read into this concept log.
+     * This method performs this operation whilst making a deep clone of the cached concepts to ensure transactions
+     * do not accidentally break the central ontology cache.
+     *
+     * @param conceptLog The thread bound concept log to read the snapshot into.
+     */
+    private void loadOntologyCacheIntoTransactionCache(ConceptLog conceptLog){
+        ConcurrentMap<TypeName, TypeImpl> cachedOntologySnapshot = getCachedOntology().asMap();
+
+        //Read central cache into conceptLog cloning only base concepts. Sets clones later
+        for (TypeImpl type : cachedOntologySnapshot.values()) {
+            conceptLog.cacheConcept(clone(type));
+        }
+
+        //Iterate through cached clones completing the cloning process.
+        //This part has to be done in a separate iteration otherwise we will infinitely recurse trying to clone everything
+        for (TypeImpl<?, ?> type : getCloneCache().values()) {
+            TypeImpl centralType = cachedOntologySnapshot.get(type.getName());
+
+            if(type.isRelationType()){
+                ((RelationTypeImpl) type).copyCachedConcepts((RelationTypeImpl) centralType);
+            } else if(type.isRoleType()){
+                ((RoleTypeImpl) type).copyCachedConcepts((RoleTypeImpl) centralType);
+            } else {
+                type.copyCachedConcepts(centralType);
+            }
+        }
+
+        //Purge clone cache to save memory
+        localCloneCache.remove();
+    }
+
+    /**
+     *
+     * @param type The type to clone
+     * @param <X> The type of the concept
+     * @return The newly deep cloned set
+     */
+    @SuppressWarnings({"unchecked", "ConstantConditions"})
+    <X extends Type> X clone(X type){
+        //Is a cloning even needed?
+        if(getCloneCache().containsKey(type.getName())){
+            return (X) getCloneCache().get(type.getName());
+        }
+
+        //If the clone has not happened then make a new one
+        TypeImpl clonedType = null;
+        if(type.isRoleType()) {
+            clonedType = new RoleTypeImpl((RoleTypeImpl) type);
+        } else if(type.isRelationType()) {
+            clonedType = new RelationTypeImpl((RelationTypeImpl) type);
+        } else if(type.isEntityType()) {
+            clonedType = new EntityTypeImpl((EntityTypeImpl) type);
+        } else if(type.isRuleType()) {
+            clonedType = new RuleTypeImpl((RuleTypeImpl) type);
+        } else if(type.isResourceType()) {
+            clonedType = new ResourceTypeImpl((ResourceTypeImpl) type);
+        } else if(type.isType()) {
+            clonedType = new TypeImpl((TypeImpl) type);
+        }
```

Comment:
Since I am travelling would you mind letting me do this in another PR since its more of a code cleanup than a bug or issue? 

I want to get this is quickly so we can find issues with it faster.